blog/2021/01/snoopi_deep.md

```diff
@@ -0,0 +1,221 @@
+   ├─ InferenceTimingNode: 0.000071/0.000071 on InferenceFrameInfo for MyType(::Int64) with 0 direct children
+   └─ InferenceTimingNode: 0.000122/0.000453 on InferenceFrameInfo for Main.SnoopDemo.dostuff(::MyType{Int64}) with 2 direct children
+      ├─ InferenceTimingNode: 0.000083/0.000161 on InferenceFrameInfo for Main.SnoopDemo.extract(::MyType{Int64}) with 2 direct children
+      │  ├─ InferenceTimingNode: 0.000045/0.000045 on InferenceFrameInfo for getproperty(::MyType{Int64}, ::Symbol) with 0 direct children
+      │  └─ InferenceTimingNode: 0.000034/0.000034 on InferenceFrameInfo for getproperty(::MyType{Int64}, x::Symbol) with 0 direct children
+      └─ InferenceTimingNode: 0.000170/0.000170 on InferenceFrameInfo for Main.SnoopDemo.domath(::Int64) with 0 direct children
+```
+
+This tree structure reveals the caller-callee relationships, showing the specific types that were used for each `MethodInstance`.
+Indeed, as the calls to `getproperty` reveal, it goes beyond the types and even shows the results of [constant propagation](https://en.wikipedia.org/wiki/Constant_folding);
+the `getproperty(::MyType{Int64}, x::Symbol)` (note `x::Symbol` instead of just plain `::Symbol`) means that the call was `getproperty(y, :x)`, which corresponds to `y.x` in the definition of `extract`.
+
+@@note
+**Box 3** Generally we speak of [call graphs](https://en.wikipedia.org/wiki/Call_graph) rather than call trees.
+But because inference results are cached (a.k.a., we only "visit" each node once), we obtain a tree as a depth-first-search of the full call graph.
+@@
+
+Each node in this tree is accompanied by a pair of numbers.
+The first number is the *exclusive* inference time (in seconds), meaning the time spent inferring the particular MethodInstance, not including the time spent inferring its callees.
+The second number is the *inclusive* time, which is the exclusive time plus the time spent on the callees.
+Therefore, the inclusive time is always at least as large as the exclusive time.
+
+The `ROOT` node is a bit different: its exclusive time measures the time spent on all operations *except* inference.
+In this case, we see that the entire call took approximately 10ms, of which 9.3ms was spent on activities besides inference.
+Almost all of that was code-generation, but it also includes the time needed to run the code.
+Just 0.76ms was needed to run type-inference on this entire series of calls.
+As you will quickly discover as you use `@snoopi_deep`, inference takes much more time on more complicated code.
+
+You can extract the `MethodInstance` with
+
+```
+julia> Core.MethodInstance(tinf)
+MethodInstance for ROOT()
+
+julia> Core.MethodInstance(tinf.children[1])
+MethodInstance for packintype(::Int64)
+```
+
+## Visualizing the output
+
+We can also display this tree as a flame graph, using the [ProfileView](https://github.com/timholy/ProfileView.jl) package:
```

Comment:
```suggestion
We can also display this tree as a flame graph, using either the [ProfileView](https://github.com/timholy/ProfileView.jl) or [PProf](https://github.com/JuliaPerf/PProf.jl) packages:
```
😉 

We've been using _both_ PProf and ProfileView together when we've been investigating. ProfileView because it shows the profile in order, and PProf because it is built specifically for handling _very large_ profiles, and some of our profiles are... woof... huge:
<img width="1216" alt="Screen Shot 2020-12-18 at 9 58 40 AM" src="https://user-images.githubusercontent.com/1582097/105248075-cd494680-5b43-11eb-986b-978b7ce73c7d.png">

But the ability to search and filter and highlight and change through different views all together make it _super_ valuable for exploring very large codebases:
<img width="1439" alt="Screen Shot 2020-12-14 at 5 06 47 PM" src="https://user-images.githubusercontent.com/1582097/105248197-fbc72180-5b43-11eb-9485-b8d0c8d3be0c.png">

<img width="1552" alt="blurred2" src="https://user-images.githubusercontent.com/1582097/105249019-688eeb80-5b45-11eb-9c97-48c70fc879f2.png">
(hehe sorry i blurred it)


Maybe we can even include an example here of working with PProf and discuss the pros and cons of each tool? :)

Reply:
I'd be happy to have this mentioned! I'd rely on you to put that in, just because I don't have your experience.